src/vcml/models/dma/pl330.cpp

```diff
@@ -0,0 +1,1144 @@
+    if (!dma->read_queue.empty(ch->tag)) {
+        ch->set_state(pl330::channel::AT_BARRIER);
+        ch->stall = 1;
+        return;
+    } else {
+        ch->set_state(pl330::channel::EXECUTING);
+    }
+}
+
+static void pl330_insn_dmasev(pl330* dma, pl330::channel* ch, u8 opcode,
+                              u8* args, int len) {
+    u8 ev_id;
+
+    if (args[0] & 0b0111) {
+        pl330_handle_ch_fault(*dma, *ch, pl330::channel::OPERAND_INVALID);
+        return;
+    }
+    ev_id = (args[0] >> 3) & 0b0001'1111;
+    if (ev_id >= dma->cr0.get_field<CR0_NUM_EVENTS>()) {
+        pl330_handle_ch_fault(*dma, *ch, pl330::channel::OPERAND_INVALID);
+        return;
+    }
+    if (ch->csr & CNS && !(dma->cr3 & (1 << ev_id))) {
+        pl330_handle_ch_fault(*dma, *ch, pl330::channel::CH_EVNT_ERR);
+        return;
+    }
+    if (dma->inten & (1 << ev_id)) {
+        dma->intmis |= (1 << ev_id);
+        //  trace_pl330_dmasev_evirq(ev_id);
+        dma->irq[ev_id] = true;
+    }
+    //    std::cout<< "pc: " << ch->cpc << " pl330_insn_dmasev event id: " <<
+    //    (u32)ev_id << std::endl; trace_pl330_dmasev_event(ev_id);
+    dma->int_event_ris |= (1 << ev_id);
+}
+
+static void pl330_mn_insn_dmasev(pl330* dma, pl330::channel* ch, u8 opcode,
+                                 u8* args, int len) {
+    u8 ev_id;
+
+    if (args[0] & 0b0111) {
+        pl330_handle_mn_fault(*dma, pl330::manager::OPERAND_INVALID);
+        return;
+    }
+    ev_id = (args[0] >> 3) & 0b0001'1111;
+    if (ev_id >= dma->cr0.get_field<CR0_NUM_EVENTS>()) {
+        pl330_handle_mn_fault(*dma, pl330::manager::OPERAND_INVALID);
+        return;
+    }
+    if (dma->manager.dsr & DNS && !(dma->cr3 & (1 << ev_id))) {
+        pl330_handle_mn_fault(*dma, pl330::manager::EVNT_ERR);
+        return;
+    }
+    if (dma->inten & (1 << ev_id)) {
+        dma->intmis |= (1 << ev_id);
+        //  trace_pl330_insn_dmasev_evirq(ev_id);
+        dma->irq[ev_id] = true;
+    }
+    std::cout << "pl330_mn_insn_dmasev" << std::endl;
+    //    trace_pl330_insn_dmasev_event(ev_id);
+    dma->int_event_ris |= (1 << ev_id);
+}
+
+static void pl330_insn_dmast(pl330* dma, pl330::channel* ch, u8 opcode,
+                             u8* args, int len) {
+    u8 bs = opcode & 0b11;
+
+    if (bs == 0b10) {
+        pl330_handle_ch_fault(*dma, *ch, pl330::channel::OPERAND_INVALID);
+        return;
+    }
+    if ((bs == 0b01 && ch->request_flag == request_type::BURST) ||
+        (bs == 0b11 && ch->request_flag == request_type::SINGLE)) {
+        /* Perform NOP */
+        return;
+    }
+
+    u32 num = ch->ccr.get_field<CCR_DST_BURST_LEN>() + 1;
+    u32 size = (u32)1 << ch->ccr.get_field<CCR_DST_BURST_SIZE>();
+    bool inc = ch->ccr & ccr_bits::DST_INC;
+    ch->stall = !dma->write_queue.push(
+        pl330::queue_entry{ ch->dar, size, num, inc, 0, ch->tag });
+    if (!ch->stall) {
+        // std::cout << "pc: " << ch->cpc << " pl330_insn_dmast to: " <<
```

Comment:
Please check commented out code. The model should also never write directly to `cout`